L52-objc-bazel-support.md

```diff
@@ -0,0 +1,124 @@
+gRPC Objective-C Bazel Build Support
+----
+* Author: tonyzhehaolu
+* Approver: mxyan
+* Status: Draft
+* Implemented in: Bazel and Starlark
+* Last updated: Today's date
+* Discussion at: ???
+
+## Abstract
+
+Proposes a set of Bazel rules for building iOS applications with gRPC Objective-C library.
+
+
+## Background
+
+The gRPC Objective-C library so far only supports installation via Cocoapods. Requests for Bazel support are continually raised.
+
+In addition to the available native rules (`objc_library` and `proto_library`), `objc_proto_lib` and `objc_grpc_lib` need to be created because the native `objc_proto_library` is actually [not usable](https://github.com/bazelbuild/bazel/issues/7348). Some other rules are also needed in order to compile the actual library.
+
+### Related Proposals
+`objc_proto_lib` and `objc_grpc_lib` are built upon the implementation of the `generate_cc` rule defined in [generate_cc.bzl](https://github.com/grpc/grpc/blob/bazel_test/bazel/generate_cc.bzl). There doesn't seem to be a proposal for that, though.
+
+
+## Proposal
+
+For now, assume that the WORKSPACE root is the gRPC repository.
+
+### Dependency Graph
+
+According to the dependencies in an iOS application that uses gRPC Objc library (shown below), we created two `objc_library` rules in the `src/objective-c` package within the `com_github_grpc_grpc` workspace: `grpc_objc` and `grpc_objc_protoRPC`.
+
+![gRPC Objective-C Library Dependency](L52_graphics/dependency.png)
+
+* Rule `//src/objective-c:grpc_objc` compiles all the files in `GRPCClient/` and `RxLibrary/`. It is dependent on a ready rule `//:grpc` which compiles core gRPC. It is publicly visible so that any application-specific Objective-C code can depend on it. It is only necessary when the app does not use protocol buffers (which means the service stub libraries, plus all that they are dependent on, are not included) - it is a rare case.
+* `//src/objective-c:grpc_protoRPC` does the `ProtoRPC/` directory. It is also made publicly visible so that the generated service stubs can be compiled depending on this rule. Users do not need to manually add this label to `deps`, though.
+* The Objective-C stubs are generated *and compiled* into native Bazel `objc_library`'s via `objc_proto_lib` and `objc_grpc_lib`. Details about these two custom rules are discussed in the upcoming sections.
```

Comment:
How about renaming `objc_grpc_lib` to `objc_grpc_library` to follow the rule name convention?

And `objc_proto_lib` to `objc_proto_grpc_lib`? (we don't want to mess with the name `objc_proto_library` because bazel team may release it one day)

I am not totally sure whether the suggestion makes sense though. Open to discussion.

Reply:
I named it `objc_proto_lib` because it does the same thing as what `objc_proto_library` is supposed to while not having the exact same name. And `objc_grpc_lib` was to contrast `objc_proto_lib` because it only generates service stubs. But if it's better to stick with the rule name convention then I can adopt your suggestions.